### PR TITLE
remove axis properties

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**
The `axes` attribute assigned dimension properties such as sparsity or insertedness, which is required for certain simplifications. Setting these properties in every constructor is however very expensive. This PR effectively makes the attribute lazy by querying the inverse operation (`_desparsify`, `_uninsert`) and testing for a nonzero return value.

### CODE REVIEW

File: nutils/evaluable.py [hunk 1]
```diff
@@ -3239,6 +3356,16 @@ def _desparsify(self, axis):
...
 
+  def _uninsert(self, axis):
+    if axis < self.ndim - 2:
+      uninserted = self.func._uninsert(axis)
+      if uninserted is not None:
+        return Unravel(uninserted, *self.shape[-2:])
+    else:
+      uninserted = self.func._uninsert(self.ndim-2)
+      if uninserted is not None:
+        return InsertAxis(uninserted, self.shape[-2 if axis == self.ndim-1 else -1])
```
Comment (joostvanzwieten) [inline nutils/evaluable.py:3313]: I think the `else` block is unnecessary because of `InsertAxis._unravel`.

File: nutils/evaluable.py [hunk 2]
```diff
@@ -1832,13 +1827,14 @@ def _loopsum(self, index):
...
-      func1, func2 = func2, func1
-    assert isinstance(func1._axes[axis], Sparse)
-    return [(ind, multiply(f1, _take(func2, ind, axis))) for ind, f1 in func1._desparsify(axis)]
+    desparsified = func1._desparsify(axis)
+    if desparsified is None:
+      desparsified = func2._desparsify(axis)
+      if desparsified is None:
+        return
+      func2 = func1
+    return [(ind, multiply(f1, _take(func2, ind, axis))) for ind, f1 in desparsified]
```
Comment (joostvanzwieten) [inline nutils/evaluable.py:1787]: Suggestion:

```python
funcs = tuple(self.funcs):
for func1, func2 in funcs, reversed(funcs):
  desparsified1 = func1._desparsify(axis)
  if desparsified1 is not None:
    return [(ind, multiply(f1, _take(func2, ind, axis))) for ind, f1 in desparsified1]
```

Comment (gertjanvanzwieten) [reply nutils/evaluable.py:1787]: done.

File: nutils/evaluable.py [hunk 3]
```diff
@@ -3206,8 +3278,7 @@ def _multiply(self, other):
     return Ravel(Multiply([self.func, Unravel(other, *self.func.shape[-2:])]))
 
   def _add(self, other):
-    if isinstance(other, Ravel) and equalshape(other.func.shape[-2:], self.func.shape[-2:]):
-      return Ravel(Add([self.func, other.func]))
+    return Ravel(self.func + Unravel(other, *self.func.shape[-2:]))
```
Comment (joostvanzwieten) [inline nutils/evaluable.py:3075]: This seems to be unrelated to patch 'add _inflations'.

File: nutils/evaluable.py [hunk 4]
```diff
@@ -67,8 +67,9 @@ def simplified(value):
 def asarray(arg):
   try:
     return arg.as_evaluable_array
-  except AttributeError:
-    pass
+  except AttributeError as e:
+    if 'as_evaluable_array' not in str(e):
+      raise
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:72]: removed mine

File: nutils/evaluable.py [hunk 5]
```diff
@@ -1376,6 +1376,8 @@ def _transpose(self, axes):
 
   def _takediag(self, axis1, axis2):
     assert axis1 < axis2
+    if axis1 == self.ndim-2 and self.axes[:-2] == tuple(range(self.ndim-2)):
+      return TakeDiag(self.func)
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:1394]: done.

File: nutils/evaluable.py [hunk 6]
```diff
@@ -751,6 +751,68 @@ def swapaxes(arg, axis1, axis2):
...
+
+  Given one or more equally shaped array arguments, return the shortest common
+  axis vector along with function arguments such that the original arrays can
+  be recovered by :func:`align`.
+  '''
+
+  assert args
+  if len(args) == 1:
+    return args[0]._unaligned
+  assert all(arg.ndim == args[0].ndim for arg in args[1:])
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:799]: fixed.

File: nutils/evaluable.py [hunk 7]
```diff
@@ -3715,11 +3765,14 @@ def _simplified(self):
...
-    if isinstance(self.func._axes[-1], Inserted) and self.index not in self.func.shape[-1].arguments and not equalindex(self.func.shape[-1], 1):
-      return Ravel(InsertAxis(loop_concatenate(InsertAxis(self.func._uninsert(self.func.ndim-1), 1), self.index), self.func.shape[-1]))
+    unaligned, where = unalign(self.func)
+    for iaxis, length in enumerate(self.func.shape):
+      if iaxis not in where and self.index not in length.arguments:
+        uninserted = align(unaligned, [i-(i>iaxis) for i in where], self.func.shape[:iaxis]+self.func.shape[iaxis+1:])
+        if iaxis != self.ndim-1:
+          return insertaxis(loop_concatenate(uninserted, self.index), iaxis, length)
+        elif not equalindex(length, 1):
+          return Ravel(InsertAxis(loop_concatenate(InsertAxis(uninserted, 1), self.index), length))
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:3620]: It was not entirely trivial but I came up with the following:
```python
unaligned, where = unalign(self.func)
if self.ndim-1 not in where:
  # reinsert concatenation axis, at unit length if possible so we can
  # insert the remainder outside of the loop
  unaligned = InsertAxis(unaligned, self.func.shape[-1] if self.index in self.func.shape[-1].arguments else 1)
  where += self.ndim-1,
elif where[-1] != self.ndim-1:
  # bring concatenation axis to the end
  unaligned = Transpose(unaligned, numpy.argsort(where))
  where = tuple(sorted(where))
f = loop_concatenate(unaligned, self.index)
if not equalindex(self.shape[-1], f.shape[-1]):
  # last axis was reinserted at unit length AND it was not unit length
  # originally - if it was unit length originally then we proceed only if
  # there are other insertions to promote, otherwise we'd get a recursion.
  f = Ravel(InsertAxis(f, self.func.shape[-1]))
elif len(where) == self.ndim:
  return
return align(f, where, self.shape)
```

File: nutils/evaluable.py [hunk 8]
```diff
@@ -1747,16 +1753,14 @@ def _simplified(self):
     unaligned1, unaligned2, where = unalign(func1, func2)
     if len(where) != self.ndim:
       return align(unaligned1 * unaligned2, where, self.shape)
-    diagonals = {}
+    for axes in func1._diagonals + func2._diagonals:
+      axis1, axis2 = sorted(axes)[:2]
+      return diagonalize(Multiply(takediag(func, axis1, axis2) for func in self.funcs), axis1, axis2)
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:1677]: done.

File: nutils/evaluable.py [hunk 9]
```diff
@@ -2904,6 +2917,18 @@ def _axes(self):
...
+  @property
+  def _diagonals(self):
+    diagonals = []
+    myaxes = frozenset([self.ndim-2, self.ndim-1])
+    for axes in self.func._diagonals:
+      if axes & myaxes:
+        myaxes |= axes
+      else:
+        diagonals.append(axes)
+    diagonals.append(myaxes)
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:2809]: fixed

File: nutils/evaluable.py [hunk 10]
```diff
@@ -1872,11 +1883,20 @@ def _desparsify(self, axis):
...
+        for *indices2, values2 in uninserted2._assparse:
+          indices = [None] * self.ndim
+          for i, j in enumerate(where1):
+            indices[j] = appendaxes(indices1[i], values2.shape)
+          for i, j in enumerate(where2):
+            indices[j] = prependaxes(indices2[i], values1.shape)
+          assert all(indices)
+          values = appendaxes(values1, values2.shape) * prependaxes(values2, values1.shape)
+          sparse.append((*indices, values))
+      return tuple(sparse)
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:1797]: fixed

File: nutils/evaluable.py (hunk 3 above)
Comment (gertjanvanzwieten) [reply nutils/evaluable.py:3075]: fixed

**Review by joostvanzwieten — APPROVED** [review]
